rust/src/util/types.rs

```diff
@@ -1,55 +1,325 @@
+// This is needed because of https://github.com/rust-lang/rust/issues/59878
+impl<T: Sized> IntoIterator for fil_Array<T> {
+    type Item = T;
+    type IntoIter = std::vec::IntoIter<T>;
 
-#[no_mangle]
-pub unsafe extern "C" fn fil_destroy_gpu_device_response(ptr: *mut fil_GpuDeviceResponse) {
-    let _ = Box::from_raw(ptr);
+    #[inline]
+    fn into_iter(self) -> Self::IntoIter {
+        Vec::from(self.into_boxed_slice()).into_iter()
+    }
+}
+
+/// Owned, fixed size and heap allocated array of bytes.
+#[allow(non_camel_case_types)]
+pub type fil_Bytes = fil_Array<u8>;
+
+impl<T> From<Vec<T>> for fil_Array<T> {
+    fn from(buf: Vec<T>) -> Self {
+        if buf.is_empty() {
+            return Default::default();
+        }
+        buf.into_boxed_slice().into()
+    }
+}
+
+impl<T: Clone> From<&[T]> for fil_Array<T> {
+    fn from(buf: &[T]) -> Self {
+        if buf.is_empty() {
+            return Default::default();
+        }
+        buf.to_vec().into_boxed_slice().into()
+    }
+}
+
+impl<T> From<Box<[T]>> for fil_Array<T> {
+    fn from(buf: Box<[T]>) -> Self {
+        if buf.is_empty() {
+            return Default::default();
+        }
+        let len = buf.len();
+        let ptr = NonNull::new(Box::into_raw(buf).cast());
+
+        Self { ptr, len }
+    }
+}
+
+impl<T> Drop for fil_Array<T> {
+    fn drop(&mut self) {
+        if let Some(ptr) = self.ptr {
+            let _ = unsafe { Box::from_raw(ptr.as_ptr()) };
+        }
+    }
+}
+
+impl<T> Deref for fil_Array<T> {
+    type Target = [T];
+    fn deref(&self) -> &[T] {
+        match self.ptr {
+            None => unsafe { std::slice::from_raw_parts(ptr::NonNull::dangling().as_ptr(), 0) },
+            Some(ptr) => unsafe { std::slice::from_raw_parts(ptr.as_ptr().cast(), self.len) },
+        }
+    }
+}
+
+impl From<&str> for fil_Bytes {
+    fn from(s: &str) -> Self {
+        if s.is_empty() {
+            return Default::default();
+        }
+        Box::<str>::from(s).into()
+    }
+}
+
+impl From<String> for fil_Bytes {
+    fn from(s: String) -> Self {
+        if s.is_empty() {
+            return Default::default();
+        }
+        s.into_boxed_str().into()
+    }
+}
+
+impl From<PathBuf> for fil_Bytes {
+    fn from(s: PathBuf) -> Self {
+        s.to_string_lossy().as_ref().into()
+    }
+}
+
+impl From<&Path> for fil_Bytes {
+    fn from(s: &Path) -> Self {
+        s.to_string_lossy().as_ref().into()
+    }
+}
+
+impl From<Box<str>> for fil_Bytes {
+    fn from(s: Box<str>) -> Self {
+        if s.is_empty() {
+            return Default::default();
+        }
+        let len = s.len();
+        let ptr = NonNull::new(Box::into_raw(s).cast());
+
+        Self { ptr, len }
+    }
+}
+
+impl fil_Bytes {
+    pub fn as_str(&self) -> Result<&str, Utf8Error> {
+        std::str::from_utf8(self)
+    }
+
+    pub fn as_path(&self) -> Result<PathBuf, Utf8Error> {
+        self.as_str().map(PathBuf::from)
```

Comment:
Are we even going to pretend to support windows? If not... we can use https://doc.rust-lang.org/std/os/unix/ffi/trait.OsStrExt.html.

Reply:
If yes, let me introduce you to utf16.

Reply:
fixed